bluetooth: mesh: extend oob authentication size till 32 bytes

alxelax opened the pull request:
PR includes:
Updates maximum OOB authentication size from 8 bytes till 32 bytes according to specification errata ES-27446 (mandatory changes). Since previous OOB API does not allow exposing OOB values with such width, the new API has been introduced: 
`int (*output_numeric)(bt_mesh_output_action_t act, uint8_t *numeric, size_t size);`
`int bt_mesh_input_numeric(uint8_t *numeric, size_t size);`
The previous API has been hidden under BT_MESH_REDUCED_MAX_OOB_SIZE option and left for backward compatibility with existing code base.
Additionally, PR includes updates of the relevant bsim tests, adaptation of the qualification test environment and adaptation of some samples, adaptation advertisement stack size dependency as well as size. The issues with size have been figured out during mesh_shell usage on `nrf52840dk/nrf52840` platform for manual testing oob related changes.

PavelVPV commented on include/zephyr/bluetooth/mesh/main.h:
```diff
@@ -184,6 +185,20 @@ struct bt_mesh_prov {
 	 *  @return Zero on success or negative error code otherwise
 	 */
 	int         (*output_number)(bt_mesh_output_action_t act, uint32_t num);
+#else
```
```suggestion
#else /* CONFIG_BT_MESH_REDUCED_MAX_OOB_SIZE */
```

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/provisioning.rst:
```diff
@@ -144,7 +144,7 @@ sequence should be repeated after a delay of three seconds or more.
 application receives the :c:member:`bt_mesh_prov.input` callback. The user
 response should be fed back to the Provisioning API through
-:c:func:`bt_mesh_input_string` or :c:func:`bt_mesh_input_number`. If
+:c:func:`bt_mesh_input_string` or :c:func:`bt_mesh_input_numeric`. If
```
I think we should keep reference to `bt_mesh_input_number` everywhere and also mention that it is only available if `CONFIG_BT_MESH_REDUCED_MAX_OOB_SIZE` is enabled.

PavelVPV commented on include/zephyr/bluetooth/mesh/main.h:
```diff
@@ -332,6 +348,19 @@ int bt_mesh_input_string(const char *str);
+
+/** @brief Provide provisioning input OOB numeric value.
+ *
+ *  This is intended to be called after the bt_mesh_prov input callback
```
```suggestion
 *  This is intended to be called after the @ref bt_mesh_prov.input callback
```
Same can be done for `bt_mesh_input_number`.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -360,6 +361,17 @@ config BT_MESH_OOB_AUTH_REQUIRED
+	  Related API for OOB output\input was changed and became incompatible with the previous one.
+	  Enable this option to enable the previous API and support the initial
+	  maximum OOB authentication size. Option is intended for backward compatibility.
+	  Not recommended for new designs.
```
What I don't like here is the usage of words `initial` and `later`. When this `initial` and when this `later` happened in time this paragraph doesn't explain, which doesn't make the picture clear for me as a user. Neither it explains the consequences of using old API other new one.

I think we can mention here that prior the Mesh Specification v1.1 inclusively, the maximum OOB size was 8 bytes.

What about this text:
```
By default, the maximum OOB authentication size is 32 bytes.

Enable this option to use the legacy API which is only compatible with
specification v1.1 and earlier. This limits OOB authentication to 8
bytes maximum.

This option is intended for backward compatibility.
Not recommended for new designs due to reduced security.
```

PavelVPV commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -122,57 +124,166 @@ static int check_input_auth(bt_mesh_input_action_t input, uint8_t size)
+	memset(bt_mesh_prov_link.auth, 0, auth_size - *size);
+}
+
+void bt_mesh_binary_to_ascii_digits(const uint8_t *in, size_t in_len, char *out_str)
```
I'm not sure why this function needs to be part of stack code. As far as I see it is only used for printing oob data in a readable text and only used in shell and in test. I suggest to move it to shell and test code.

Also, since it is only used for printing, I think it can be simplified using `%c`. E.g. this is how logging subsystem prints hexdump: https://github.com/zephyrproject-rtos/zephyr/blob/0df9c5fb1f21df1c34daf12664d38e77e729af6d/subsys/logging/log_minimal.c#L42-L50

alxelax replied:
My initial implementation was exactly what you requested. However, Sonar checker banned it due to huge percentage of similarity. That's the truth, since the functions are identical in shell and in test. 
The checker is quite stupid in understanding purposes. Nevertheless, it banns PR. Instead, I added the single one here. I do not like it either. However, it is not a big problem since the function should be cut out by linker if this is not used.

Any better ideas are welcome.

PavelVPV replied:
We should try to suppress Sonar I think. I'm still confident that if this function doesn't belong to stack, it should not be there.

PavelVPV commented on include/zephyr/bluetooth/mesh/main.h:
```diff
@@ -173,6 +173,7 @@ struct bt_mesh_prov {
 	void         (*capabilities)(const struct bt_mesh_dev_capabilities *cap);
 
+#if defined CONFIG_BT_MESH_REDUCED_MAX_OOB_SIZE
 	/** @brief Output of a number is requested.
```
We need to add here that this callback is enabled only if `CONFIG_BT_MESH_REDUCED_MAX_OOB_SIZE` is enabled.

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -399,18 +400,55 @@ static int cmd_proxy_solicit(const struct shell *sh, size_t argc,
 #if defined(CONFIG_BT_MESH_SHELL_PROV)
+static int ascii_decimal_to_le_bytes(const char *ascii_str, uint8_t *out, size_t *out_len)
+{
+	for (const char *p = ascii_str; *p; ++p) {
```
I guess argv[1] is always NULL-terminated?
```suggestion
	for (const char *p = ascii_str; *p != '\0'; ++p) {
```
Otherwise this should be:
```suggestion
	for (const char *p = ascii_str; *p != NULL; ++p) {
```

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -399,18 +400,55 @@ static int cmd_proxy_solicit(const struct shell *sh, size_t argc,
+		}
+
+		/* Handle overflow to next byte */
+		if (carry) {
```
```suggestion
		if (carry != 0) {
```

alxelax replied:
What is the difference to what I implemented? Isn't it the same?

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -399,18 +400,55 @@ static int cmd_proxy_solicit(const struct shell *sh, size_t argc,
+	size_t result_len = 0;
 
-	val = shell_strtoul(argv[1], 10, &err);
+	err = ascii_decimal_to_le_bytes(argv[1], result, &result_len);
```
Does this code really work?

`result_len` here is 0. But AFAIS `ascii_decimal_to_le_bytes` expects `result_len` to be equal to size of `result`.

omkar3141 commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -399,18 +400,55 @@ static int cmd_proxy_solicit(const struct shell *sh, size_t argc,
+{
+	for (const char *p = ascii_str; *p; ++p) {
+		if (!isdigit((unsigned char)*p)) {
+			return -1; /* Invalid character */
```
-EINVAL?

PavelVPV commented:
Please also list the new API in this section: https://github.com/zephyrproject-rtos/zephyr/blob/8b208b0d5a358e24e57b5795e0bfaaac9fc1e38c/doc/releases/release-notes-4.4.rst?plain=1#L56-L68

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -28,6 +28,8 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(bt_mesh_prov);
 
+#define MAX_NUMERIC_OOB_BYTES 14
```
10 power 32 represents 14 bytes

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -360,6 +361,17 @@ config BT_MESH_OOB_AUTH_REQUIRED
 	bool "OOB authentication mandates to use HMAC SHA256"
 	depends on BT_MESH_ECDH_P256_HMAC_SHA256_AES_CCM
 
+config BT_MESH_REDUCED_MAX_OOB_SIZE
```
I guess since we need to have only one callback, we need the Kconfig option. But if the Kconfig option enables the old, deprecated API, it needs to be deprecated as well so that it is removed together with the old API.

To deprecate the struct parameter, I think you can `__deprecated ` before the parameter's type and `@deprecated` in the description.

alxelax commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -208,32 +323,32 @@ int bt_mesh_prov_auth(bool is_provisioner, uint8_t method, uint8_t action, uint8
-			atomic_set_bit(bt_mesh_prov_link.flags, NOTIFY_INPUT_COMPLETE);
-			get_auth_string(str, size);
-			return bt_mesh_prov->output_string(str);
+			get_auth_string(rand_bytes, size);
```
I'm not sure if I understand this. `rand_bytes` is already in maximum size. Maximum size is `PROV_IO_OOB_SIZE_MAX` according to spec.
`size` that comes to the function might be any, below the max as well as above max, that depends on vs device capability. `bt_mesh_prov_auth` checks that size gotten from capability does not exceed this device ability.
I do not see any issue in this logic. Incoming `size` is not permanent value, and we have to check it, but OOB value shall not exceed array capacity that is checked.

PavelVPV replied:
`get_auth_string` gets `size` as size of `rand_bytes` while actual size of `rand_bytes` is not what `size` variable holds.

alxelax committed:
bluetooth: mesh: improve oob size checking

Commit improves oob size checking to catch wrong size
as early as possible and prevents zero size.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
bluetooth: mesh: zeroing not used part of auth array

Commit set zero to not used part of authentication array
to prevent garbage from the previous provisioning using.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
bluetooth: mesh: prevent getting wrong authentication size

Commit moves SHA256 authentication size under SHA256 macro.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
bluetooth: mesh: adjust adv stack dependency and size

Commit removes legacy dependancy on Host crypto.
Additionally, it adapts advertiser stack size if
provisioner feature has been enabled (tested with
mesh_shell and thread analyzer).

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
bluetooth: mesh: update max oob size till 32 bytes

Commit updates maximum OOB authentication size from 8 bytes
till 32 bytes according to specification errata ES-27446.
Since previous OOB API does not allow to expose OOB values
with such width the new API has been introduced.
The previous API was deprecated and hidden under
BT_MESH_PROV_OOB_API_LEGACY option and left for backward
compatibility with existing code base.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV commented on include/zephyr/bluetooth/mesh/main.h:
```diff
@@ -331,7 +347,20 @@ int bt_mesh_input_string(const char *str);
+/** @brief Provide provisioning input OOB numeric value.
+ *
+ *  This is intended to be called after the bt_mesh_prov input callback
+ *  has been called with BT_MESH_ENTER_NUMBER as the action.
```
```suggestion
 *  has been called with @ref BT_MESH_ENTER_NUMBER as the action.
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -456,27 +458,64 @@ static struct bt_mesh_prov prov = {
 };
 
-static int output_number(bt_mesh_output_action_t action, uint32_t number)
+void bt_mesh_binary_to_ascii_digits(const uint8_t *in, size_t in_len, char *out_str)
```
```suggestion
static void binary_to_ascii_digits(const uint8_t *in, size_t in_len, char *out_str)
```

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -523,30 +565,75 @@ static void prov_reset(void)
+	uint8_t temp[PROV_IO_OOB_SIZE_MAX];
+	size_t digit_count = 0;
+
+	memcpy(temp, in, in_len);
```
Assert if `in_len > sizeof(temp)`?

PavelVPV reviewed (changes requested):
Few minor comments, otherwise LGTM

omkar3141 reviewed (approved):
Minor comments if you would like to address them. Otherwise looks ok.
